experimental/regression_suite/ireers_tools/artifacts.py

```diff
@@ -0,0 +1,244 @@
+for log_name, log_obj in logging.Logger.manager.loggerDict.items():
+    if log_name.startswith("azure"):
+        logging.getLogger(log_name).setLevel(logging.WARNING)
+
+
+def show_progress(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize is not None:
+            t.total = tsize
+        t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+
+    return update_to
+
+
+@functools.cache
+def get_artifact_root_dir() -> Path:
+    root_path = os.getenv("IREE_TEST_FILES", default=str(Path.cwd())) + "/artifacts"
+    return Path(os.path.expanduser(root_path)).resolve()
+
+
+class ArtifactGroup:
+    """A group of artifacts with a persistent location on disk."""
+
+    _INSTANCES: Dict[str, "ArtifactGroup"] = {}
+
+    def __init__(self, group_name: str):
+        self.group_name = group_name
+        if group_name:
+            self.directory = get_artifact_root_dir() / group_name
+        else:
+            self.directory = get_artifact_root_dir()
+        self.directory.mkdir(parents=True, exist_ok=True)
+
+    @classmethod
+    def get(cls, group: Union["ArtifactGroup", str]) -> "ArtifactGroup":
+        if isinstance(group, ArtifactGroup):
+            return group
+        try:
+            return cls._INSTANCES[group]
+        except KeyError:
+            instance = ArtifactGroup(group)
+            cls._INSTANCES[group] = instance
+            return instance
+
+
+class Artifact:
+    """Some form of artifact materialized to disk."""
+
+    def __init__(
+        self,
+        group: Union[ArtifactGroup, str],
+        name: str,
+        depends: Collection["Artifact"] = (),
+    ):
+        self.group = ArtifactGroup.get(group)
+        self.name = name
+        self.depends = tuple(depends)
+
+    @property
+    def path(self) -> Path:
+        return self.group.directory / self.name
+
+    def join(self):
+        """Waits for the artifact to become available."""
+        pass
+
+    def __str__(self):
+        return str(self.path)
+
+
+class ProducedArtifact(Artifact):
+    def __init__(
+        self,
+        group: Union[ArtifactGroup, str],
+        name: str,
+        callback: Callable[["ProducedArtifact"], Any],
+        *,
+        always_produce: bool = False,
+        depends: Collection["Artifact"] = (),
+    ):
+        self.group = ArtifactGroup.get(group)
+        super().__init__(group, name, depends)
+        self.name = name
+        self.callback = callback
+        self.always_produce = always_produce
+
+    @property
+    def stamp_path(self) -> Path:
+        """Path of a stamp file which indicates successful transfer."""
+        return self.path.with_suffix(self.path.suffix + ".stamp")
+
+    def start(self) -> "ProducedArtifact":
+        if not self.always_produce and self.stamp_path.exists():
+            if self.path.exists():
+                print(f"Not producing {self} because it has already been produced")
+                return self
```

Comment:
> Also didn't realize, but ireers here already has a process of stamping here to check if a file is already produced. I think we have to remove this because it will skip even if there is a newer version of the file available and there's really no point when downloading to a cache because once it's there, it is never removed so not a valuable signal.

How is this caching .vmfb files? We should be recompiling after changes to the compiler, but this looks like it just used an existing .vmfb on disk:
https://github.com/iree-org/iree/actions/runs/9908082489/job/27373608457?pr=17883#step:11:37

Reply:
Yeah it is not caching properly. I am going to remove the whole stamp checking thing that is here and stick with hash checking